pwa/src/offline-interface/offline-interface.js

```diff
@@ -58,44 +59,88 @@ export class OfflineInterface {
             return null
         }
 
-        function onUpdate(registration) {
-            if (!promptUpdate) {
+        // This event emitter helps coordinate with service worker messages
+        this.offlineEvents = new EventEmitter()
+
+        // Receives messages from service worker and forwards to event emitter
+        const handleSWMessage = event => {
+            if (!event.data) {
                 return
             }
+            const { type, payload } = event.data
+            this.offlineEvents.emit(type, payload)
+        }
+        navigator.serviceWorker.addEventListener('message', handleSWMessage)
 
-            const message = i18n.t("There's an update available for this app.")
-            const onConfirm = () =>
-                registration.waiting.postMessage({
-                    type: swMsgs.skipWaiting,
-                })
+        /**
+         * Requests clients info from the active service worker. Works for
+         * both first activation and SW update by using `reg.active` worker.
+         * Arrow function to preserve `this`.
+         */
+        const getClientsInfo = registration => {
+            return new Promise(resolve => {
+                registration.active.postMessage({ type: swMsgs.getClientsInfo })
+                this.offlineEvents.once(swMsgs.clientsInfo, resolve)
+                // Clean up unused listeners eventually
+                setTimeout(() => {
+                    this.offlineEvents.removeAllListeners(swMsgs.clientsInfo)
+                }, 10000)
+            })
+        }
+
+        function getPromptOptions({ onConfirm, clientsCount }) {
+            const promptMessage = i18n.t(
+                "There's an update available for this app."
+            )
+            const confirmReloadMessage = i18n.t(
+                'There are {{n}} instances of this app open currently. Activating updates will reload all open instances, and unsaved data will be lost. Ready to reload now?',
+                { n: clientsCount }
+            )
+            const onClick = () => {
+                if (clientsCount <= 1) {
+                    // One client - go ahead and reload
+                    onConfirm()
+                } else if (window.confirm(confirmReloadMessage)) {
+                    // Multiple clients - make sure to confirm reload to avoid
+                    // data loss
+                    onConfirm()
+                }
```

Comment:
`window.confirm` used here because using React UI components here is complicated -- another alert could be used through the app runtime, but I'm also thinking about a way to refactor much of the update logic to live in React land but use the offline interface for communication; for example a `<PWAUpdateManager />` component in the app adapter